### PR TITLE
Add support for soundcloud comments

### PR DESCRIPTION
**Pull request by B0pol**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API. (just update extractor version)

By the way:
+ I fixed by the way upload date wrong on soundcloud using API v2: it now displays the good time (we were not taking on board timezone.
+ I reorganized soundcloud with the good packages (following youtube and peertube)
+ I renamed methods in CommentsInfoItem to match the method names across NPE

### CODE REVIEW

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        final CommentsInfoItemsCollector collector = new CommentsInfoItemsCollector(getServiceId());
+        collectStreamsFrom(collector, json.getArray("collection"));
+
+        return new InfoItemsPage<>(collector, getNextPageUrl());
+    }
+
+    @Override
+    public void onFetchPage(@Nonnull Downloader downloader) throws IOException, ExtractionException {
+        Response response = downloader.get(getUrl());
+        System.err.println(response.responseBody());
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java:63]: Remove the `System.err.println`s

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/SoundcloudService.java [hunk 2]
```diff
@@ -10,18 +10,22 @@
...
+import org.schabi.newpipe.extractor.services.soundcloud.linkHandler.*;
 import org.schabi.newpipe.extractor.stream.StreamExtractor;
 import org.schabi.newpipe.extractor.subscription.SubscriptionExtractor;
 
 import java.util.List;
 
+import static java.util.Arrays.asList;
 import static java.util.Collections.singletonList;
 import static org.schabi.newpipe.extractor.StreamingService.ServiceInfo.MediaCapability.AUDIO;
+import static org.schabi.newpipe.extractor.StreamingService.ServiceInfo.MediaCapability.COMMENTS;
```
[the remark this answers is not in the document]
Comment (B0pol) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/SoundcloudService.java]: removed in the last commit

File: extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItemExtractor.java [hunk 3]
```diff
@@ -3,17 +3,44 @@
...
+    int getLikeCount() throws ParsingException;
+
+    /**
+     * The text of the comment
+     */
+    String getCommentText() throws ParsingException;
+
+    /**
+     * The upload date given by the service, unmodified
+     * see {@link StreamExtractor#getTextualUploadDate()}
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItemExtractor.java:30]: Use `@see`

File: extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItemExtractor.java [hunk 4]
```diff
@@ -3,17 +3,44 @@
...
 
 public interface CommentsInfoItemExtractor extends InfoItemExtractor {
-    String getCommentId() throws ParsingException;
-    String getCommentText() throws ParsingException;
-    String getAuthorName() throws ParsingException;
-    String getAuthorThumbnail() throws ParsingException;
+
+    /**
+     * AuthorEndpoint, in other words, link to authors' channel page
+     */
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItemExtractor.java:14]: Imho these should be renamed to `getUploaderUrl()`, `getUploaderName()`, and `getUploaderAvatarUrl()` to be consistent with the other `Extractor`s.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            json = JsonParser.object().from(response.responseBody());
+        } catch (JsonParserException e) {
+            throw new ParsingException("Could not parse json", e);
+        }
+    }
+
+    @Nonnull
+    @Override
+    public String getName() throws ParsingException {
+        return "SoundCloud comments of track " + getId();
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java:73]: Just return "Comments" like in PeerTube (and YouTube in my `Page` PR), as that function is just pointless. Maybe even do that in `CommentsExtractor` directly and remove the function from all the comments extractors.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    private void collectStreamsFrom(final CommentsInfoItemsCollector collector, final JsonArray entries) throws ParsingException {
+        String url = getUrl();
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java:77]: `final`

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/linkHandler/SoundcloudCommentsLinkHandlerFactory.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    public String getUrl(String id, List<String> contentFilter, String sortFilter) throws ParsingException {
+        try {
+            return "https://api-v2.soundcloud.com/tracks/" + id + "/comments" + "?client_id=" + clientId() +
+                    "&threaded=0" + "&filter_replies=1"; // anything but 1 = sort by new
+            // + "&limit=NUMBER_OF_ITEMS_PER_REQUEST". We let the API control (default = 10)
+            // + "&offset=OFFSET". We let the API control (default = 0, then we use nextPageUrl)
+        } catch (ExtractionException | IOException e) {
+            throw new ParsingException("Could not get comments");
+        }
+    }
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/linkHandler/SoundcloudCommentsLinkHandlerFactory.java:30]: `getUrl()` should never return an API URL.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeSubscriptionExtractor.java [hunk 8]
```diff
@@ -103,7 +103,7 @@ private String readFromInputStream(InputStream inputStream) throws InvalidSource
             }
         } catch (InvalidSourceException e) {
             throw e;
-        } catch (Throwable e) {
+        }   catch (Throwable e) {
```
Comment (wb9688) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeSubscriptionExtractor.java:106]: Why this indentation? Plus as you said, you shouldn't catch `Throwable`, and this file isn't even related to your PR

Comment (B0pol) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeSubscriptionExtractor.java:106]: No, I was just browsing with vscode and sent it on irc for bad practise. I'll remove it.

**Pull request merged**